indimail-2.5/clearopensmtp.8
.TH clearopensmtp 8
.SH NAME
.PP
clearopensmtp \- remove stale IP's authorized for pop before SMTP
.PP
updaterules \- Rebuild tcp.smtp.cdb with list of static ip stored in tcprules

.SH SYNOPSYS
.PP
.B clearopensmtp
.PP
.B updaterules

.SH DESCRIPTION
\fBclearopensmtp\fR removes old IP's added after pop authentication. These IP's are maintained
in the table relay by default. IndiMail can be configured to allow POP3/IMAP users IP's to be
added to the list of IP's which are allowed to relay through the smtp server. This is done
with the --enable-roaming-users=y option while configuring IndiMail before compilation. With
this option, users IP's who authenticate via POP3/IMAP are added to the list of IP's which can
relay through the smtp server. \fBqmail-smtpd(8)\fR must be run with \fBtcpserver(1)\fR
\fB\-x\fR option. If the host is a clustered domain, \fBclearopensmtp\fR uses the DBINFO
structure to cycle through all the mailstores and clear the relay tables.

\fBupdaterules\fR rebuilds the cdb (tcp.smtp.cdb) file with the list of static IP's stored in
tcprules (\fBtcp.smtp\fR) file and the list of IP's in \fBopen-smtp\fR or relay table. This
rebuilding is needed only if IndiMail's \fBqmail-smtpd(8)\fR has not been installed, and one is
running some other version of smtpd daemon. \fBupdaterules\fR may also be needed if one is
running a version of smtpd on some foreign hosts (e.g. Relay servers) which do not have access
to IndiMail's authentication databases. IndiMail's \fBqmail-smtpd(8)\fR gets the authenticated
user list at run time from the \fBrelay\fR table and hence does not require the IP's in the
\fBtcp.smtp.cdb\fR file.

Each authenticated pop users IP is added with a time stamp. Every time \fBclearopensmtp\fR is
run, this list is checked for time stamps which are older than the 
\fB\-\-enable\-relay\-clear\-minutes\fR option. The default is 30 minutes. Any IP with a time
stamp older than this number are removed from the list. 

.SH OPTIONS
.PP
These options are for \fBclearopensmtp\fR only.
.TP
\fB\-C\fR
Connect to cluster. Allows \fBclearopensmtp\fR to operate on all hosts defined
in \fBdbinfo\fR Mail Cluster Definition (MCD) File. This option causes \fIrelay\fR table
to be cleared on all hosts defined in the MCD file.

.TP
\fB\-t\fR
Clear the table \fIrelay\fB in MySQL only.

.TP
\fB\-s\fR
Rebuild the cdb only.

.SH RETURN VALUE
0 if all steps were successful, non-zero otherwise.


indimail-2.5/clearopensmtp.c
/*
 * $Log: clearopensmtp.c,v $
 * Revision 2.13  2017-03-13 13:37:29+05:30  Cprogrammer
 * replaced qmaildir with sysconfdir
 *
 * Revision 2.12  2016-05-17 17:09:39+05:30  Cprogrammer
 * use control directory set by configure
 *
 * Revision 2.11  2011-11-09 19:42:23+05:30  Cprogrammer
 * removed getversion
 *
 * Revision 2.10  2010-05-05 14:40:51+05:30  Cprogrammer
 * added connect_all argument to vclear_open_smtp
 *
 * Revision 2.9  2008-05-28 21:34:27+05:30  Cprogrammer
 * added option to selectively run clearing of relay table or run update tcp.smtp.cdb
 *
 * Revision 2.8  2008-05-28 16:33:57+05:30  Cprogrammer
 * removed USE_MYSQL, removed cdb code
 *
 * Revision 2.7  2005-12-29 22:40:16+05:30  Cprogrammer
 * use getEnvConfigStr to set variables from environment variables
 *
 * Revision 2.6  2004-10-27 14:55:34+05:30  Cprogrammer
 * close mysql before exit
 *
 * Revision 2.5  2002-08-25 22:48:48+05:30  Cprogrammer
 * made control dir configurable
 *
 * Revision 2.4  2002-08-11 00:25:05+05:30  Cprogrammer
 * added deletion of PID file
 *
 * Revision 2.3  2002-07-07 22:28:13+05:30  Cprogrammer
 * changed name of variable clear_minutes to clear_seconds
 *
 * Revision 2.2  2002-06-26 03:17:17+05:30  Cprogrammer
 * correction for non-distributed code
 *
 * Revision 2.1  2002-05-13 12:35:42+05:30  Cprogrammer
 * removed argument mytime to vclear_open_smtp()
 *
 * Revision 1.4  2001-12-12 19:24:04+05:30  Cprogrammer
 * added is_alread_running() to prevent multiple instances to run
 *
 * Revision 1.3  2001-11-24 12:17:53+05:30  Cprogrammer
 * version information added
 *
 * Revision 1.2  2001-11-20 10:53:39+05:30  Cprogrammer
 * *** empty log message ***
 *
 * Revision 1.1  2001-10-24 18:14:53+05:30  Cprogrammer
 * Initial revision
 *
 */
#include "indimail.h"

#ifndef	lint
static char     sccsid[] = "$Id: clearopensmtp.c,v 2.13 2017-03-13 13:37:29+05:30 Cprogrammer Stab mbhangui $";
#endif

#ifdef POP_AUTH_OPEN_RELAY
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <string.h>

void            usage();

int
main(int argc, char **argv)
{
	char           *mcdfile, *controldir, *tmpstr;
	char            TmpBuf[MAX_BUFF];
	time_t          clear_seconds;
	int             c, errflag, job_type = 1, cluster_opt = 0;

	errflag = 0;
	while (!errflag && (c = getopt(argc, argv, "ts")) != -1)
	{
		switch (c)
		{
		case 'C':
			cluster_opt = 1;
			break;
		case 't':
			job_type = 2;
			break;
		case 's':
			job_type = 3;
			break;
		default:
			errflag = 1;
			break;
		}
	}
	if (errflag > 0)
	{
		usage();
		return(1);
	}
	if (is_already_running("clearopensmtp"))
	{
		fprintf(stderr, "clearopensmtp is already running\n");
		return(1);
	}
	if (job_type == 1 || job_type == 3)
	{
		clear_seconds = atoi(RELAY_CLEAR_MINUTES) * 60;
		errflag = vclear_open_smtp(clear_seconds, cluster_opt);
#ifdef CLUSTERED_SITE
		close_db();
#else
		vclose();
#endif
	}
	if (job_type == 1 || job_type == 2)
	{
		getEnvConfigStr(&controldir, "CONTROLDIR", CONTROLDIR);
		getEnvConfigStr(&mcdfile, "MCDFILE", MCDFILE);
		if (*mcdfile == '/')
			scopy(TmpBuf, mcdfile, MAX_BUFF);
		else {
			if (*controldir == '/')
				snprintf(TmpBuf, MAX_BUFF, "%s/%s", controldir, mcdfile);
			else {
				getEnvConfigStr(&tmpstr, "SYSCONFDIR", SYSCONFDIR);
				snprintf(TmpBuf, MAX_BUFF, "%s/%s/%s", tmpstr, controldir, mcdfile);
			}
		}
		if (access(TmpBuf, F_OK))
		{
			if (job_type == 2 && vauth_open((char *) 0))
				return(1);
			errflag = update_rules(1);
			vclose();
		}
		unlink("/tmp/clearopensmtp.PID");
	}
	return (errflag);
}

void
usage()
{
	char           *relay_table;

	getEnvConfigStr(&relay_table, "RELAY_TABLE", RELAY_DEFAULT_TABLE);
	printf("usage: clearopensmtp [options]\n");
	printf("options: -V (print version number)\n");
	printf("         -t (run updaterules)\n");
	printf("         -s (clear MySQL RELAY TABLE (%s))\n", relay_table);
}
#else
int
main()
{
	printf("IndiMail not configured with --enable-roaming-users=y\n");
	return(0);
}
#endif

void
getversion_clearopensmtp_c()
{
	printf("%s\n", sccsid);
	printf("%s\n", sccsidh);
}
